presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestTransformCorrelatedScalarSubquery.java

```diff
@@ -0,0 +1,189 @@
+                                        markDistinct(
+                                                "is_distinct",
+                                                ImmutableList.of("unique"),
+                                                lateral(
+                                                        ImmutableList.of("corr"),
+                                                        assignUniqueId(
+                                                                "unique",
+                                                                values("corr")),
+                                                        filter(
+                                                                "1 = a",
+                                                                values("a")))))));
+    }
+
+    @Test
+    public void rewritesOnSubqueryWithProjection()
+    {
+        tester().assertThat(rule)
+                .on(p -> p.lateral(
+                        ImmutableList.of(p.symbol("corr")),
+                        p.values(p.symbol("corr")),
+                        p.enforceSingleRow(
+                                p.project(
+                                        Assignments.of(p.symbol("a2"), p.expression("a * 2")),
+                                        p.filter(
+                                                p.expression("1 = a"), // TODO use correlated predicate, it requires support for correlated subqueries in plan matchers
+                                                p.values(p.symbol("a")))))))
+                .matches(
+                        project(
+                                filter(
+                                        ensureScalarSubquery(),
+                                        markDistinct(
+                                                "is_distinct",
+                                                ImmutableList.of("unique"),
+                                                lateral(
+                                                        ImmutableList.of("corr"),
+                                                        assignUniqueId(
+                                                                "unique",
+                                                                values("corr")),
+                                                        project(ImmutableMap.of("a2", expression("a * 2")),
+                                                                filter("1 = a",
+                                                                        values("a"))))))));
+    }
+
+    @Test
+    public void rewritesOnSubqueryWithProjectionOnTopEnforceSingleNode()
+    {
+        tester().assertThat(rule)
+                .on(p -> p.lateral(
+                        ImmutableList.of(p.symbol("corr")),
+                        p.values(p.symbol("corr")),
+                        p.project(Assignments.of(p.symbol("a3"), p.expression("a2 + 1")),
+                                p.enforceSingleRow(
```

Comment:
add a test where there is scalar aggregation in right subtree

Reply:
So I read your comment. The point here is to have correlated subtree that produces at most single row to ensure no additional nodes (mark distinct) are added then, but `values` should do too

Reply:
Yes, but this rule does not check if is subquery is at most scalar. This will be handled before this rule is fired.

We can talk offline about it and change it later if needed.